src/gui/SampleThumbnail.cpp

```diff
@@ -0,0 +1,156 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleThumbnail.h"
+
+#include <QDebug>
+#include <QFileInfo>
+
+namespace lmms {
+
+SampleThumbnail::Thumbnail::Thumbnail(std::vector<Peak> peaks, double samplesPerPeak)
+	: m_peaks(std::move(peaks))
+	, m_samplesPerPeak(samplesPerPeak)
+{
+}
+
+SampleThumbnail::Thumbnail::Thumbnail(const SampleFrame* buffer, size_t size, size_t width)
+	: m_peaks(width)
+	, m_samplesPerPeak(static_cast<double>(size) / width)
+{
+	for (auto peakIndex = std::size_t{0}; peakIndex < width; ++peakIndex)
+	{
+		const auto beginSample = buffer + static_cast<size_t>(std::floor(peakIndex * m_samplesPerPeak));
+		const auto endSample = buffer + static_cast<size_t>(std::ceil((peakIndex + 1) * m_samplesPerPeak));
+		const auto [min, max] = std::minmax_element(&beginSample->left(), &endSample->left());
+		m_peaks[peakIndex] = Peak{*min, *max};
+	}
+}
+
+SampleThumbnail::Thumbnail SampleThumbnail::Thumbnail::zoomOut(float factor, size_t from, size_t to) const
+{
+	assert(factor >= 1 && "Invalid zoom out factor");
+
+	auto peaks = std::vector<Peak>((to - from) / factor);
+	for (auto peakIndex = std::size_t{0}; peakIndex < peaks.size(); ++peakIndex)
+	{
+		const auto beginAggregationAt = m_peaks.begin() + from + static_cast<size_t>(std::floor(peakIndex * factor));
+		const auto endAggregationAt = m_peaks.begin() + from + static_cast<size_t>(std::ceil((peakIndex + 1) * factor));
+		peaks[peakIndex] = std::accumulate(beginAggregationAt, endAggregationAt, Peak{});
+	}
+
+	return Thumbnail{std::move(peaks), m_samplesPerPeak * factor};
+}
+
+SampleThumbnail::Thumbnail SampleThumbnail::Thumbnail::zoomOut(float factor) const
+{
+	return zoomOut(factor, 0, m_peaks.size());
+}
+
+SampleThumbnail::SampleThumbnail(const Sample& sample)
+{
+	auto entry = SampleThumbnailEntry{sample.sampleFile(), QFileInfo{sample.sampleFile()}.lastModified()};
+	if (!entry.filePath.isEmpty())
+	{
+		const auto it = s_sampleThumbnailCacheMap.find(entry);
+		if (it != s_sampleThumbnailCacheMap.end())
+		{
+			m_thumbnailCache = it->second;
+			return;
+		}
+
+		if (s_sampleThumbnailCacheMap.size() == MaxSampleThumbnailCacheSize)
+		{
+			const auto leastUsed = std::min_element(s_sampleThumbnailCacheMap.begin(), s_sampleThumbnailCacheMap.end(),
+				[](const auto& a, const auto& b) { return a.second.use_count() < b.second.use_count(); });
+			s_sampleThumbnailCacheMap.erase(leastUsed->first);
+		}
+
+		s_sampleThumbnailCacheMap[std::move(entry)] = m_thumbnailCache;
```

Comment:
The `s_sampleThumbnailCacheMap` accesses aren't thread-safe. Could that be a problem?

Reply:
No, not at all. Visualization only happens on the UI thread.